internal/backend_s3.go

```diff
@@ -725,53 +746,127 @@ func (s *S3Backend) CopyBlob(param *CopyBlobInput) (*CopyBlobOutput, error) {
+	timeStampISO8601Format := time.Now().Format("20060102T150405Z")
+	timestampYMD := time.Now().Format("20060102")
+	//payload := strings.NewReader("<file contents here>") // just keeping as this as generated by
+	payload := strings.NewReader("")
+	// Generate hash for payload (in our case currently empty)
+	hasher := sha256.New()
+	hasher.Write([]byte(""))
+	hashedPayload := hex.EncodeToString(hasher.Sum(nil)) // should be e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+
+	// combine host and filepath to get full url string
+	req, err := http.NewRequest(method, "https://"+host+filePath, payload)
 
-	req, resp := s.GetObjectRequest(&get)
-	err := req.Send()
 	if err != nil {
-		return nil, mapAwsError(err)
-	}
+		fmt.Println(err)
+	}
+	signature := generateSignature(timeStampISO8601Format, timestampYMD, hashedPayload, host, filePath, method)
+	req.Header.Add("X-Amz-Content-Sha256", hashedPayload)
+	req.Header.Add("X-Amz-Date", timeStampISO8601Format)
+	req.Header.Add("Authorization", "AWS4-HMAC-SHA256 Credential="+os.Getenv("AWS_ACCESS_KEY_ID")+"/"+timestampYMD+
+		"/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date,Signature="+signature)
+	// Example of what the request header should look like below
+	// AWS4-HMAC-SHA256 Credential=AKIAIOSFODNN7EXAMPLE/20130524/us-east-1/s3/aws4_request,SignedHeaders=host;range;x-amz-content-sha256;x-amz-date,Signature=f0e8bdb87c964420e857bd35b5d6ed310bd44f0170aba48dd91039c6036bdb41
+	return req
+}
 
+// s.bucket and param.Key combine nicely with good slash management. The final path we want looks something like;
+// /bucket/path/path2/.../file.txt
+// Example -> s.bucket: 1121045215484495542 and param.Key: jose/new,file.txt
+func returnURIPath(fullPath string) string {
+	pathToClean := strings.Split(fullPath, `/`)
+	cleanedPath := ""
+	for i := range pathToClean {
+		cleanedPath += "/" + url.QueryEscape(pathToClean[i])
+	}
+	// Must replace any `+`'s to %20 https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+	// the QueryEscape encodes the space to +
+	cleanedPath = strings.ReplaceAll(cleanedPath, "+", "%20")
+	return cleanedPath
+}
+func (s *S3Backend) GetBlob(param *GetBlobInput) (*GetBlobOutput, error) {
+	cleanedPath := returnURIPath(s.bucket + param.Key)
+	request := createRequest(os.Getenv("BUCKET_HOST"), "GET", cleanedPath)
+	res, e := s.httpClient.Do(request)
+	if e != nil {
+		s3Log.Debugf(e.Error())
+	}
+	// Build the information to be sent in the response
+	etag := res.Header.Get("ETag")
+	lastModified, _ := time.Parse("Mon, 02 Jan 2006 15:04:05 GMT", res.Header.Get("Last-Modified"))
+	size, _ := strconv.ParseUint(res.Header.Get("ContentLength"), 10, 64)
+	storageClass := res.Header.Get("x-amz-storage-class")
+	contentType := res.Header.Get("Content-Type")
+	amzRequest := res.Header.Get("x-amz-request-id") + ": " + res.Header.Get("x-amz-id-2")
+	amzMeta := make(map[string]*string)
+	for key, val := range res.Header {
```

Comment:
Can we avoid nested For? same as above comment

Reply:
The first loop is for going through a list of all the headers returned because we need the headers with `x-amz-meta-`. The second is for building the header itself and is needed to build the `amzMeta` header that plays with the pre-defined go types that goofys has, example;
![image](https://github.com/user-attachments/assets/252269e0-73a7-47eb-b5b1-7b8804109e36)